### PR TITLE
[#264] cxx node builder

### PR DESCRIPTION
**Pull request by elfenpiff**
## Pre-Review Checklist for the PR Author

1. [x] Add sensible notes for the reviewer
1. [x] PR title is short, expressive and meaningful
1. [x] Relevant issues are linked in the [References](#references) section
1. [x] Every source code file has a copyright header with `SPDX-License-Identifier: Apache-2.0 OR MIT`
1. [x] Branch follows the naming format (`iox2-123-introduce-posix-ipc-example`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`[#123] Add posix ipc example`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Assign PR to reviewer
1. [x] All checks have passed (except `task-list-completed`)

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx2/blob/main/doc/release-notes/iceoryx2-unreleased.md

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Unit tests have been written for new behavior
- [x] Public API is documented
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

Relates to #264

### CODE REVIEW

File: iceoryx2-ffi/cxx/src/node_details.cpp [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+// https://www.apache.org/licenses/LICENSE-2.0, or the MIT license
+// which is available at https://opensource.org/licenses/MIT.
+//
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+#include "iox2/node_details.hpp"
+
+namespace iox2 {
+NodeDetails::NodeDetails(NodeName name, const Config& config)
+    : m_node_name { std::move(name) }
```
Comment (elfenpiff) [inline iceoryx2-ffi/cxx/src/node_details.cpp:17]: I know it looks weird but it seems to be the new best practice, see: https://clang.llvm.org/extra/clang-tidy/checks/modernize/pass-by-value.html

**Review by elBoberido — APPROVED** [review]
Looks good. Only some remarks for future functionality and Duration might be more efficient to pass by value.

File: iceoryx2-ffi/cxx/include/iox2/node.hpp [hunk 2]
```diff
@@ -38,52 +39,75 @@ enum class NodeEvent : uint8_t {
...
+    auto operator=(Node&&) noexcept -> Node&;
     Node(const Node&) = delete;
     auto operator=(const Node&) -> Node& = delete;
+    ~Node();
+
+    /// Returns the name of the node inside a [`NodeNameView`].
+    auto name() const -> NodeNameView;
+
+    /// Returns the unique id of the [`Node`].
+    auto id() const -> NodeId;
```
[the remark this answers is not in the document]
Comment (elfenpiff) [reply iceoryx2-ffi/cxx/include/iox2/node.hpp:59]: It has a size of 16 bytes so I would keep it as a copy.

File: iceoryx2-ffi/cxx/include/iox2/node.hpp [hunk 3]
```diff
@@ -38,52 +39,75 @@ enum class NodeEvent : uint8_t {
...
-        -> iox::expected<void, NodeListFailure> {
-        IOX_TODO();
-    }
+    /// Returns the [`ServiceBuilder`] to create a new service. The
+    /// [`ServiceName`] of the [`Service`] is provided as argument.
+    auto service_builder(const ServiceName& name) const -> ServiceBuilder<T>;
+
+    /// Waits a given time for a [`NodeEvent`]. If the [`cycle_time`] has passed
+    /// [`NodeEvent::Tick`] is returned.
+    auto wait(const iox::units::Duration& cycle_time) const -> NodeEvent;
```
[the remark this answers is not in the document]
Comment (elfenpiff) [reply iceoryx2-ffi/cxx/include/iox2/node.hpp:67]: Addressed in a follow up PR.

**Pull request merged**